src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1386,6 +1413,37 @@ public IList<string> ValidValues
         {
             get
             {
+                if (validValuesGeneratorType != null)
+                {
+                    ValidValuesCacheEntry validValuesCacheEntry;
+                    if (validValuesCache.TryGetValue(validValuesGeneratorType, out validValuesCacheEntry))
+                    {
+                        var currentTime = DateTime.Now;
+                        if (DateTime.Compare(validValuesCacheEntry.cacheEntryCreationTime.AddSeconds(CacheExpiration), currentTime) < 0)
+                        {
+                            // Update expired Entry.
+                            validValuesCacheEntry.validValues = validValuesCacheEntry.validValuesGenerator.GetValidValues()?.ToArray();
+                            validValuesCacheEntry.cacheEntryCreationTime = currentTime;
+                            validValuesCache[validValuesGeneratorType] = validValuesCacheEntry;
+                        }
+                    }
+                    else
+                    {
+                        var validValuesGenerator = (IValidateSetValuesGenerator)Activator.CreateInstance(validValuesGeneratorType);
+                        var validValues = validValuesGenerator.GetValidValues()?.ToArray();
+                        var cacheEntryCreationTime = DateTime.Now;
+                        validValuesCacheEntry = new ValidValuesCacheEntry(validValuesGenerator, validValues, cacheEntryCreationTime);
+                        validValuesCache.TryAdd(validValuesGeneratorType, validValuesCacheEntry);
+                    }
+                    if (validValuesCacheEntry.validValues == null)
+                    {
+                        throw new ValidationMetadataException(
+                            "ValidateSetGeneratedValidValuesListIsEmpty",
+                            null,
+                            Metadata.ValidateSetGeneratedValidValuesListIsEmpty);
+                    }
+                    _validValues = validValuesCacheEntry.validValues;
```

Comment:
Sure, you could add `CacheExpiration` to `IValidateSetValuesGenerator`, but it would be something extra that some generators have no need for.

I agree we should clean up `validValuesCache` - and ideally, I think that cache is optionally available to generators but not required. This is why I suggested an abstract base class that generators could derive from.